tensorboard/plugins/projector/vz_projector/data.ts

```diff
@@ -370,6 +365,91 @@ export class DataSet {
+  async projectUmap(nComponents: number, nNeighbors: number, stepCallback: (iter: number) => void) {
+    this.hasUmapRun = true;
+    this.umap = new UMAP({nComponents, nNeighbors});
+
+    let currentEpoch = 0;
+    const epochStepSize = 10;
+    const sampledIndices = this.shuffledDataIndices.slice(0, UMAP_SAMPLE_SIZE);
+  
+    const sampledData = sampledIndices.map(i => this.points[i]);
+    // TODO: Switch to a Float32-based UMAP internal
+    const X = sampledData.map(x => Array.from(x.vector));
+
+    this.nearest = await this.computeKnn(sampledData, nNeighbors);
+    
+    const nEpochs = await util.runAsyncTask('Initializing UMAP...', () => {
+      const knnIndices = this.nearest.map(row => row.map(entry => entry.index));
+      const knnDistances = this.nearest.map(row => row.map(entry => entry.dist));
+
+      // Initialize UMAP and return the number of epochs.
+      return this.umap.initializeFit(X, knnIndices, knnDistances);
+    }, UMAP_MSG_ID);
+    
+    // Now, iterate through all epoch batches of the UMAP optimization, updating
+    // the modal window with the progress rather than animating each step since
+    // the UMAP animation is not nearly as informative as t-SNE.
+    return new Promise((resolve, reject) => {
+      const step = () => {
+        // Compute a batch of epochs since we don't want to update the UI
+        // on every epoch.
+        const epochsBatch = Math.min(epochStepSize, nEpochs - currentEpoch);
+        for (let i = 0; i < epochsBatch; i++) {
+          currentEpoch = this.umap.step();
+        }
+        const progressMsg = `Optimizing UMAP (epoch ${currentEpoch} of ${nEpochs})`;
+        
+        // Wrap the logic in a util.runAsyncTask in order to correctly update
+        // the modal with the progress of the optimization.
+        util.runAsyncTask(progressMsg, () => {
+          if (currentEpoch < nEpochs) {
+            requestAnimationFrame(step);
+          } else {
+            const result = this.umap.getEmbedding();
+            sampledIndices.forEach((index, i) => {
+               const dataPoint = this.points[index];
+    
+              dataPoint.projections['umap-0'] = result[i][0];
+              dataPoint.projections['umap-1'] = result[i][1];
+              if (nComponents === 3) {
+                dataPoint.projections['umap-2'] = result[i][2];
+              }
+            });
+            this.projections['umap'] = true;
+    
+            logging.setModalMessage(null, UMAP_MSG_ID);
+            this.hasUmapRun = true;
+            stepCallback(currentEpoch);
+            resolve();
+          }
+        }, UMAP_MSG_ID, 0).catch(error => {
+          logging.setModalMessage(null, UMAP_MSG_ID);
+          reject(error);
+        });
+      }
+
+      requestAnimationFrame(step);
+    });
+  }  
+
+  /** Computes KNN to provide to the UMAP and t-SNE algorithms. */
+  private async computeKnn(data: DataPoint[], nNeighbors: number): Promise<knn.NearestEntry[][]> {
```

Comment:
nit: Please wrap at 80 like the rest of this file; Prettier suggests:

```typescript
  private async computeKnn(
    data: DataPoint[],
    nNeighbors: number,
  ): Promise<knn.NearestEntry[][]> {
    /* … */
  }
```

Reply:
I'm basically useless without prettier.